// next@15.0.0-rc.0/dist/server/async-storage/request-async-storage-wrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "RequestAsyncStorageWrapper", {
    enumerable: true,
    get: function() {
        return RequestAsyncStorageWrapper;
    }
});
const _approuterheaders = require("../../client/components/app-router-headers");
const _headers = require("../web/spec-extension/adapters/headers");
const _requestcookies = require("../web/spec-extension/adapters/request-cookies");
const _cookies = require("../web/spec-extension/cookies");
const _draftmodeprovider = require("./draft-mode-provider");
const _utils = require("../web/utils");
const _aftercontext = require("../after/after-context");
function getHeaders(headers) {
    const cleaned = _headers.HeadersAdapter.from(headers);
    for (const param of _approuterheaders.FLIGHT_PARAMETERS){
        cleaned.delete(param.toString().toLowerCase());
    }
    return _headers.HeadersAdapter.seal(cleaned);
}
function getMutableCookies(headers, onUpdateCookies) {
    const cookies = new _cookies.RequestCookies(_headers.HeadersAdapter.from(headers));
    return _requestcookies.MutableRequestCookiesAdapter.wrap(cookies, onUpdateCookies);
}
const RequestAsyncStorageWrapper = {
    /**
   * Wrap the callback with the given store so it can access the underlying
   * store using hooks.
   *
   * @param storage underlying storage object returned by the module
   * @param context context to seed the store
   * @param callback function to call within the scope of the context
   * @returns the result returned by the callback
   */ wrap (storage, { req, res, renderOpts }, callback) {
        let previewProps = undefined;
        if (renderOpts && 'previewProps' in renderOpts) {
            // TODO: investigate why previewProps isn't on RenderOpts
            previewProps = renderOpts.previewProps;
        }
        const [wrapWithAfter, afterContext] = createAfterWrapper(renderOpts);
        function defaultOnUpdateCookies(cookies) {
            if (res) {
                res.setHeader('Set-Cookie', cookies);
            }
        }
        const cache = {};
        const store = {
            get headers () {
                if (!cache.headers) {
                    // Seal the headers object that'll freeze out any methods that could
                    // mutate the underlying data.
                    cache.headers = getHeaders(req.headers);
                }
                return cache.headers;
            },
            get cookies () {
                if (!cache.cookies) {
                    // if middleware is setting cookie(s), then include those in
                    // the initial cached cookies so they can be read in render
                    const requestCookies = new _cookies.RequestCookies(_headers.HeadersAdapter.from(req.headers));
                    if ('x-middleware-set-cookie' in req.headers && typeof req.headers['x-middleware-set-cookie'] === 'string') {
                        const setCookieValue = req.headers['x-middleware-set-cookie'];
                        const responseHeaders = new Headers();
                        for (const cookie of (0, _utils.splitCookiesString)(setCookieValue)){
                            responseHeaders.append('set-cookie', cookie);
                        }
                        const responseCookies = new _cookies.ResponseCookies(responseHeaders);
                        // Transfer cookies from ResponseCookies to RequestCookies
                        for (const cookie of responseCookies.getAll()){
                            requestCookies.set(cookie.name, cookie.value ?? '');
                        }
                    }
                    // Seal the cookies object that'll freeze out any methods that could
                    // mutate the underlying data.
                    cache.cookies = _requestcookies.RequestCookiesAdapter.seal(requestCookies);
                }
                return cache.cookies;
            },
            get mutableCookies () {
                if (!cache.mutableCookies) {
                    cache.mutableCookies = getMutableCookies(req.headers, (renderOpts == null ? void 0 : renderOpts.onUpdateCookies) || (res ? defaultOnUpdateCookies : undefined));
                }
                return cache.mutableCookies;
            },
            get draftMode () {
                if (!cache.draftMode) {
                    cache.draftMode = new _draftmodeprovider.DraftModeProvider(previewProps, req, this.cookies, this.mutableCookies);
                }
                return cache.draftMode;
            },
            reactLoadableManifest: (renderOpts == null ? void 0 : renderOpts.reactLoadableManifest) || {},
            assetPrefix: (renderOpts == null ? void 0 : renderOpts.assetPrefix) || '',
            afterContext
        };
        return wrapWithAfter(store, ()=>storage.run(store, callback, store));
    }
};
function createAfterWrapper(renderOpts) {
    var _renderOpts_experimental, _renderOpts_ComponentMod;
    const isAfterEnabled = (renderOpts == null ? void 0 : (_renderOpts_experimental = renderOpts.experimental) == null ? void 0 : _renderOpts_experimental.after) ?? false;
    if (!renderOpts || !isAfterEnabled) {
        return [
            (_, callback)=>callback(),
            undefined
        ];
    }
    const { waitUntil, onClose } = renderOpts;
    const cacheScope = (_renderOpts_ComponentMod = renderOpts.ComponentMod) == null ? void 0 : _renderOpts_ComponentMod.createCacheScope();
    const afterContext = (0, _aftercontext.createAfterContext)({
        waitUntil,
        onClose,
        cacheScope
    });
    const wrap = (requestStore, callback)=>afterContext.run(requestStore, callback);
    return [
        wrap,
        afterContext
    ];
}
